File: packages/catcher-http-ts/src/sse/router.ts
/**
 * SSE Line Router — routes raw SSE text lines.
 *
 * Silent (consumed by library):
 *   - empty line → event separator
 *   - `:` prefix → comment / heartbeat
 *   - `id:` → record lastEventId
 *   - `retry:` → adjust reconnect interval
 *
 * Yield (passed to user as-is, with prefix):
 *   - `data:`, `event:`, and any other content lines
 */

export type RouteAction =
  | { kind: 'yield'; line: string }
  | { kind: 'silent' }
  | { kind: 'setLastEventId'; id: string }
  | { kind: 'setRetry'; ms: number }

export function routeLine(line: string): RouteAction {
  if (line === '') return { kind: 'silent' }
  if (line.startsWith(':')) return { kind: 'silent' }
  if (line.startsWith('id:')) {
    return { kind: 'setLastEventId', id: line.slice(3).trimStart() }
  }
  if (line.startsWith('retry:')) {
    const ms = parseInt(line.slice(6).trim(), 10)
    if (Number.isFinite(ms) && ms >= 0) return { kind: 'setRetry', ms }
  }
  return { kind: 'yield', line }
}

File: packages/catcher-http/src/sse/router.rs
/// Line routing result
pub enum RouteAction {
    /// Content line — yield to user
    Yield(String),
    /// Control line — silently consumed (heartbeat / empty line)
    Silent,
    /// `id:` line — record last_event_id for reconnect
    SetLastEventId(String),
    /// `retry:` line — adjust reconnect interval
    SetRetry(u64),
}

/// Route a single SSE text line.
pub fn route_line(line: &str) -> RouteAction {
    if line.is_empty() {
        return RouteAction::Silent;
    }
    if line.starts_with(':') {
        return RouteAction::Silent;
    }
    if let Some(id) = line.strip_prefix("id:") {
        return RouteAction::SetLastEventId(id.trim_start().to_string());
    }
    if let Some(retry) = line.strip_prefix("retry:") {
        if let Ok(ms) = retry.trim().parse::<u64>() {
            return RouteAction::SetRetry(ms);
        }
    }
    // All other lines (data:, event:, etc.) are yielded as-is
    RouteAction::Yield(line.to_string())
}
